x86: add support for AWS T3 instances

bobafetthotmail opened the pull request:
Amazon AWS T3 cloud instances require kernel support
for the Elastic Fabric Adapter to access storage
and for Elastic Network Adapter to use network
interfaces.

Since these features are needed to access the
root filesystem and use default network,
enable them by default in x86_64 kernel.

Signed-off-by: Alberto Bursi <bobafetthotmail@gmail.com>

This was tested by and reported working correctly by **mirik** forum user here https://forum.openwrt.org/t/openwrt-x86-64-running-on-aws/85172/14?u=bobafetthotmail

blocktrron commented on target/linux/x86/64/config-5.4:
```diff
@@ -159,6 +159,7 @@ CONFIG_EFI_RUNTIME_WRAPPERS=y
 CONFIG_EFI_STUB=y
 # CONFIG_EFI_TEST is not set
 # CONFIG_EFI_VARS is not set
+CONFIG_ENA_ETHERNET=y
```
IMHO it is better to build this as a module and not include it into the default kernel.

ynezz replied:
Since these features are needed to access the root filesystem and use default network, enable them by default in x86_64 kernel.

bobafetthotmail replied:
if I build this as a module I would still have to add it to the default packages for x86-64 target, because as mentioned this is the default (and only) type of network interface of these cloud instances.
Would that be significantly different than just enabling it in the kernel?
In x86-64 kernel config there are a lot of various virtual/paravirtual drivers already for kvm/xen, should this be different?

blocktrron replied:
> Would that be significantly different than just enabling it in the kernel?

Yes, you can remove the module using the imagebuilder then.

ynezz replied:
Indeed, good point!

blocktrron replied:
Note that I was only referring to the network adapter, which most are already (although it being mostly physical adapters) distributed as modules.

Yes, this was done very inconsequent in the past for virtualization targets (x86 is not the most well maintained in that regard), however i wouldn't let this hold us back to opt for the more modual approach.

> Ideological dislike for Amazon perhaps?

My credit card statements do not back this assumption.

bobafetthotmail replied:
(sigh) fine, you win, I'll make a module.

Since I'm going to edit makefiles anyway, can I also move all network devices of other virtualization targets as modules (and add the modules as default) so that the situation is now consistent?

I can test only KVM and Vmware though, although just moving to kmods shouldn't affect functionality.

blocktrron commented on target/linux/x86/64/config-5.4:
```diff
@@ -311,6 +313,7 @@ CONFIG_ND_CLAIM=y
 CONFIG_NEED_DMA_MAP_STATE=y
 CONFIG_NET_FAILOVER=y
 CONFIG_NET_FLOW_LIMIT=y
+CONFIG_NET_VENDOR_AMAZON=y
```
Nevertheless, this is already enabled in the default configuration

bobafetthotmail replied:
good catch, that option enables Amazon network drivers for T3 cloud instance, so it imho should be removed from default config and kept only here.

I can add a commit to drop that from default config, is that OK?

blocktrron replied:
All vendors are disabled by default in the default configuration, as none of them change the artifacts generated but simplify the module packaging.

bobafetthotmail replied:
seem to me they are all enabled by default in default config https://github.com/openwrt/openwrt/blob/41b096b44861cf630fff9a991ebc9b4b006ce7ac/target/linux/generic/config-5.4#L3539

But anyway, if the consensus is not touching that I'm just dropping the CONFIG_NET_VENDOR_AMAZON=y  from this patch.

bobafetthotmail commented:
updated, now the network is a module and the CONFIG_NET_VENDOR_AMAZON=y is removed

hauke commented on target/linux/x86/modules.mk:
```diff
@@ -4,6 +4,23 @@
+  TITLE:=Elastic Network Adapter (for Amazon AWS T3)
+  DEPENDS:=@TARGET_x86_64 
+  KCONFIG:=CONFIG_ENA_ETHERNET
+  FILES:=$(LINUX_DIR)/drivers/platform/x86/pcengines-apuv2.ko
```
This path looks wrong to me, this should be:
```
FILES:=$(LINUX_DIR)/drivers/net/ethernet/amazon/ena/ena.ko
```

bobafetthotmail commented:
updated, fixed path for module binary

hauke commented on target/linux/x86/modules.mk:
```diff
@@ -4,6 +4,23 @@
+  DEPENDS:=@TARGET_x86_64 
+  KCONFIG:=CONFIG_ENA_ETHERNET
+  FILES:=$(LINUX_DIR)/drivers/net/ethernet/amazon/ena/ena.ko
+  AUTOLOAD:=$(call AutoLoad,12,amzn-ena)
```
This has to match the kernel module name, so ena instead of amzn-ena.
Please try if AutoProbe works instead.

Did you test this change?

bobafetthotmail replied:
The original commit where I enabled both in the kernel config were actually tested by the guy in the forum, the move to a module was not, I thought I grasped better how kmods worked, but it seems I didn't.
I'll ask him to test again after I addressed your last feedback.

hauke commented on target/linux/x86/modules.mk:
```diff
@@ -4,6 +4,23 @@
 # This is free software, licensed under the GNU General Public License v2.
 #
 
+define KernelPackage/amzn-ena
```
Please rename the package to amazon-ena

bobafetthotmail commented:
updated with latest feedback, not tested yet. Will update with test results when the guy from the forum tells me.

bobafetthotmail commented:
And the feedback is in, AutoProbe does not work, trying with AutoLoad instead

bobafetthotmail committed:
x86: add support for AWS T3 instances

Amazon AWS T3 cloud instances require kernel support
for the Elastic Fabric Adapter to access storage
and for Elastic Network Adapter to use network
interfaces.

Since the Fabric Adapter is needed to access
root filesystem, enable in x86_64 kernel.

Elastic Network Adapter goes in a module,
and add this module to default list in x86_64.
The module is set to AutoLoad because AutoProbe does
not seem to load it.

Signed-off-by: Alberto Bursi <bobafetthotmail@gmail.com>

bobafetthotmail commented:
With AutoLoad it works correctly, so updated with that

hauke commented:
Thank you for the pull request, it was added to master.

bobafetthotmail commented:
can you add this to the 21.02 branch as well?
